src/python/pants/backend/jvm/tasks/coursier_resolve.py

```diff
@@ -0,0 +1,478 @@
+from pants.backend.jvm.tasks.coursier.coursier_subsystem import CoursierSubsystem
+from pants.backend.jvm.tasks.nailgun_task import NailgunTask
+from pants.base.exceptions import TaskError
+from pants.base.fingerprint_strategy import FingerprintStrategy
+from pants.base.workunit import WorkUnit, WorkUnitLabel
+from pants.java.jar.jar_dependency_utils import M2Coordinate, ResolvedJar
+from pants.util.contextutil import temporary_file
+from pants.util.dirutil import safe_mkdir
+from pants.util.process_handler import subprocess
+
+
+logger = logging.getLogger(__name__)
+
+
+class CoursierError(Exception):
+  pass
+
+
+class CoursierMixin(NailgunTask):
+
+  @classmethod
+  def implementation_version(cls):
+    return super(CoursierMixin, cls).implementation_version() + [('CoursierMixin', 0)]
+
+  @classmethod
+  def register_options(cls, register):
+    super(CoursierMixin, cls).register_options(register)
+
+  @classmethod
+  def subsystem_dependencies(cls):
+    return super(CoursierMixin, cls).subsystem_dependencies() + (JarDependencyManagement, CoursierSubsystem)
+
+  @staticmethod
+  def _compute_jars_to_resolve_and_to_exclude(jars, artifact_set, manager):
+    """
+
+    :param jars_by_key:
+    :param artifact_set:
+    :param manager:
+    :return:
+    """
+
+    if artifact_set is None:
+      artifact_set = PinnedJarArtifactSet()
+
+    jars_by_key = OrderedDict()
+    for jar in jars:
+      jars_for_the_key = jars_by_key.setdefault((jar.org, jar.name), [])
+      jars_for_the_key.append(jar)
+
+    jars_to_resolve = []
+    exclude_args = set()
+
+    untouched_pinned_artifact = set(M2Coordinate.create(x) for x in artifact_set)
+
+    for k, jar_list in jars_by_key.items():
+      for i, dep in enumerate(jar_list):
+        direct_coord = M2Coordinate.create(dep)
+
+        if direct_coord in artifact_set:
+          managed_coord = artifact_set[direct_coord]
+          untouched_pinned_artifact.remove(managed_coord)
+
+          if direct_coord.rev != managed_coord.rev:
+            # It may be necessary to actually change the version number of the jar we want to resolve
+            # here, because overrides do not apply directly (they are exclusively transitive). This is
+            # actually a good thing, because it gives us more control over what happens.
+            coord = manager.resolve_version_conflict(managed_coord, direct_coord, force=dep.force)
+
+            # Once a version is settled, we force it anyway
+            jar_list[i] = dep.copy(rev=coord.rev, force=True)
+
+      jars_to_resolve.extend(jar_list)
+      for jar in jar_list:
+        for ex in jar.excludes:
+          # `--` means exclude. See --soft-exclude-file in `coursier fetch --help`
+          ex_arg = "{}:{}--{}:{}".format(jar.org, jar.name, ex.org, ex.name)
+          exclude_args.add(ex_arg)
+
+    return jars_to_resolve, exclude_args, untouched_pinned_artifact
+
+  @property
+  def cache_target_dirs(self):
```

Comment:
I'm pretty sure that this will not work, as Coursier/Ivy are places where the entire context (rather than individual targets) matter. That makes them candidates for "manual" caching: https://www.pantsbuild.org/dev_tasks.html#enabling-artifact-caching-for-tasks

Also, you'd need to be very careful doing this in a mixin, because the thing you're mixing into might not want things cached.

Reply:
removed. no caching is used. only valid and not valid for coursier resolve.